RavenDB-19629 Atomic guard is out of sync if it was deleted in a regular tx

ayende opened the pull request:
When we attempt to load a document in a cluster wide transaction, and it is missing, we now also fetch the associated atomic guard.
That enables us to handle scenarios where the deletion happened outside of a cluster wide transaction.
When we save the "new" document, we'll attach to it the existing index (missing) that we had for the atomic guard, so everything works.


### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19629/Atomic-guard-is-out-of-sync-if-it-was-deleted-in-a-regular-tx


### Additional description

Now when a cluster wide transaction is loading a document that was deleted in a normal transaction, it'll get back its atomic guard (that was orphaned) and be able to proceed from there.

### Type of change

- Bug fix

### How risky is the change?

- Low 
- 
### Backward compatibility

This is both backward & forward compatible. 
Older servers on newer clients will just ignore this (resulting in the same behavior)
Older clients on newer servers will not trigger this

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

ayende committed:
RavenDB-19629 When we attempt to load a document in a cluster wide transaction, and it is missing, we now also fetch the associated atomic guard.
That enables us to handle scenarios where the deletion happened outside of a cluster wide transaction.
When we save the "new" document, we'll attach to it the existing index (missing) that we had for the atomic guard, so everything works.

karmeli87 commented:
LGTM, can you please check whether your changes also handle this case https://issues.hibernatingrhinos.com/issue/RavenDB-16881 and does it make sense to handle them both in this PR?

ayende commented:
Okay, that is a bit of a problem.
Technically speaking, easy enough to do, but I realized that we may have a race here.

Consider the scenario where we have a sequence of cluster-wide and non cluster-wide operations.
Such that the atomic guard is out of sync with the change vector.

We then do a read on the documents storage and the compare exchange.
*However,* think about the sequence of events.
We first commit the cluster wide transaction, then we apply the transaction.
It is possible that you'll have a "partially committed" cluster-wide transaction (such that compare exchange already exists) and the document hasn't been updated yet.

What happens then?

@karmeli87

ayende committed:
RavenDB-16881 Mixing cluster tx with normal tx might out sync the guard

Added checks that ensures that when we read the guard from the cluster storage, we ensure that we are at least as synced as that.

karmeli87 commented on src/Raven.Server/Documents/Handlers/DocumentHandler.cs:
```diff
@@ -236,13 +238,39 @@ private async Task GetDocumentsByIdAsync(DocumentsOperationContext context, Micr
+                        long? guardIndex = includeCompareExchangeValues.GetAtomicGuardIndex(ClusterTransactionCommand.GetAtomicGuardKey(id), lastModifiedIndex);
+                        if (guardIndex != null)
+                        {
+                            var list = document.ChangeVector.ToChangeVectorList();
+                            list.Add(new ChangeVectorEntry
+                            {
+                                Etag = guardIndex.Value,
+                                DbId = Database.ClusterTransactionId,
+                                NodeTag = ChangeVectorParser.TrxnInt
+                            });
+                            document.ChangeVector = list.SerializeVector();
+                        }
```
better to use `ChangeVectorUtils.TryUpdateChangeVector` I think

ayende replied:
Done

karmeli87 commented on src/Raven.Server/Documents/Handlers/DocumentHandler.cs:
```diff
@@ -236,13 +238,39 @@ private async Task GetDocumentsByIdAsync(DocumentsOperationContext context, Micr
+                        }
+                    }
+                    else if (loadFromClusterWideTx && 
+                             document.ChangeVector.Contains(ChangeVectorParser.TrxnTag) == false)
```
better to check the `ClusterTransactionId`. the `TrxnTag` can be from external replication

ayende replied:
Good point

karmeli87 commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -34,6 +35,21 @@ public abstract class ClusterTransactionOperationsBase
+        {
+            if (_missingDocumentsToAtomicGuardIndex?.TryGetValue(docId, out var index) == true)
+            {
+                changeVector = $"TXRN:{index}-{_clusterTransactionId}";
```
`ChangeVectorParser.TrxnTag`

ayende replied:
Cannot use that here, I"m afraid, that is on the _server_ only, not exposed to the client.

ppekrol replied:
Let's move that to Constants then as internal under Documens.ChangeVectors.TrxnTag?

ppekrol replied:
I'm a bit afraid that this moves the responsibility from server to client, I think it would be better for server to build the CV and send it?

ayende replied:
Done

karmeli87 commented on src/Raven.Server/Documents/Includes/IncludeCompareExchangeValuesCommand.cs:
```diff
@@ -63,26 +64,36 @@ internal void Gather(Document document)
+
+            if(value.Index > maxAllowedRaftId)
+                return null; // we are seeing partially committed value, skip it
+            if (value.Index < 0)
```
when can this happened?

ayende replied:
That can happen if the atomic guard does not exists

ayende committed:
RavenDB-19629 PR comments

ppekrol commented on src/Raven.Client/Documents/Commands/GetDocumentsCommand.cs:
```diff
@@ -265,5 +268,10 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
 
         public override bool IsReadRequest => true;
+
+        public void FromClusterWideTx()
```
internal

ppekrol commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -231,7 +247,17 @@ internal void RegisterCompareExchangeValues(BlittableJsonReaderObject values)
-                    RegisterCompareExchangeValue(CompareExchangeValueResultParser<BlittableJsonReaderObject>.GetSingleValue(value, materializeMetadata: false, _session.Conventions));
+                    var val = CompareExchangeValueResultParser<BlittableJsonReaderObject>.GetSingleValue(value, materializeMetadata: false, _session.Conventions);
+                    if(includingMissingAtomicGuards  &&
+                        val.Key.StartsWith(Constants.CompareExchange.RvnAtomicPrefix, StringComparison.InvariantCultureIgnoreCase))
```
OrdinalIgnoreCase?

ayende replied:
done

ppekrol commented on src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:
```diff
@@ -532,6 +558,11 @@ private static void ThrowInvalidExpiresMetadata(string message)
             }
         }
+
+        public void SetClusterTransactionId(string clusterTransactionId)
```
internal?

ayende replied:
removed

ppekrol commented on src/Raven.Client/Documents/Commands/GetDocumentsCommand.cs:
```diff
@@ -102,6 +103,8 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
                 .Append("/docs?");
 
+            if (_loadFromClusterWideTx)
+                pathBuilder.Append("&loadFromClusterWideTx=true");
```
better to pass transaction mode here? e.g. `txMode=ClusterWide`

ppekrol commented on test/SlowTests/Issues/RavenDB-19629.cs:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+    
+    [Fact]
```
RavenFact :)

ppekrol commented on test/SlowTests/Issues/RavenDB-19629.cs:
```diff
@@ -0,0 +1,82 @@
+    public RavenDB_19629(ITestOutputHelper output) : base(output)
+    {
+    }
+    [Fact]
```
RavenFact :)

ayende committed:
RavenDB-19629 Provide the required change vector for missing guards from the server side, changed query string name to be more generic

ppekrol committed:
RavenDB-19629 adjustments

ppekrol committed:
RavenDB-19629 changed attribute in the tests
